modules/javafx.graphics/src/main/java/com/sun/glass/utils/NativeLibLoader.java

```diff
@@ -150,6 +173,60 @@ private static void loadLibraryInternal(String libraryName) {
         }
     }
 
+   /**
+    * If there is a library with the platform-correct name at the
+    * root of the resources in this jar, use that.
+    */
+    private static boolean loadLibraryFromResource(String libraryName, List<String> dependencies, Class caller) {
+        try {
+            // first preload dependencies
+            if (dependencies != null) {
+                for (String dep: dependencies) {
+                    loadLibraryFromResource(dep, null, caller);
+                }
+            }
+            String reallib = "/"+libPrefix+libraryName+libSuffix;
+            InputStream is = caller.getResourceAsStream(reallib);
+            if (is != null) {
+                String fp = cacheLibrary(is, reallib);
+                System.load(fp);
+                if (verbose) {
+                    System.err.println("Loaded library " + reallib + " from resource");
+                }
+                return true;
+            }
+        } catch (Throwable t) {
+            // we should only be here if the resource exists in the module, but 
+            // for some reasons it can't be loaded. 
+            System.err.println("Loading library " + libraryName + " from resource failed: " + t);
+            t.printStackTrace();
+        }
+        return false;
+    }
+
+    private static String cacheLibrary(InputStream is, String name) throws IOException {
+        String jfxVersion = System.getProperty("javafx.version", "versionless");
+        String userCache = System.getProperty("user.home") + "/.openjfx/cache/" + jfxVersion;
+        File cacheDir = new File(userCache);
+        if (cacheDir.exists()) {
+            if (!cacheDir.isDirectory()) {
+                throw new IOException ("Cache exists but is not a directory: "+cacheDir);
+            }
+        } else {
+            if (!cacheDir.mkdirs()) {
+                throw new IOException ("Can not create cache at "+cacheDir);
+            }
+        }
+        // we have a cache directory. Add the file here
+        File f = new File(cacheDir, name);
+        // if it exists, remove. Todo: check checksum and keep if same.
```

Comment:
Yes, this will be needed in order to run multiple applications at the same time.

Reply:
If the filenames of different versions are the same, they need to be extracted in different directories though.

Reply:
Directories are already different for different versions: String userCache = System.getProperty("user.home") + "/.openjfx/cache/" + jfxVersion;

However, when 2 applications using the same version are executed together, they will share the same directory. Libraries should never differ between versions, so I start to wonder if a checksum is needed at all. If the checksum of the "new" library is different from the installed one, there is an issue. This should not happen, and it indicates that either the old install is tampered with, or the new resource is corrupt. Both causes are evil, and I don't know what the best generic solution would be?

Reply:
Yes, if you enforce it with version numbers, that should work fine too. That's what SWT does.

Reply:
But that doesn't work when running custom builds during development...

Reply:
in that case, it makes sense to use the one from the resource, and removing the old one. I'll do that.